rimage/transform/parallel_projection.go

```diff
@@ -95,5 +97,149 @@ func (pp *ParallelProjection) PointCloudToRGBD(cloud pointcloud.PointCloud) (*ri
+func (ppRM *ParallelProjectionOntoXZWithRobotMarker) PointCloudToRGBD(cloud pointcloud.PointCloud,
+) (*rimage.Image, *rimage.DepthMap, error) {
+	meta := cloud.MetaData()
+
+	var X, Z []float64
+	cloud.Iterate(0, 0, func(pt r3.Vector, data pointcloud.Data) bool {
+		X = append(X, pt.X)
+		Z = append(Z, pt.Z)
+		return true
+	})
+
+	// Calculate max and min range to be represented in the output image and, if needed, cropping it based on
+	// the mean and standard deviation of the X and Z coordinates
+	meanX, err := stats.Mean(X)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "calculation of the X-coord's mean during pcd projection failed")
+	}
+	stdevX, err := stats.StandardDeviation(X)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "calculation of the X-coord's stdev during pcd projection failed")
+	}
+
+	meanZ, err := stats.Mean(Z)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "calculation of the Z-coord's mean during pcd projection failed")
+	}
+	stdevZ, err := stats.StandardDeviation(Z)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "calculation of the Z-coord's stdev during pcd projection failed")
+	}
+
+	maxX := math.Min(meanX+float64(sigmaLevel)*stdevX, meta.MaxX)
+	minX := math.Max(meanX-float64(sigmaLevel)*stdevX, meta.MinX)
+	maxZ := math.Min(meanZ+float64(sigmaLevel)*stdevZ, meta.MaxZ)
+	minZ := math.Max(meanZ-float64(sigmaLevel)*stdevZ, meta.MinZ)
+
+	// Change the max and min values to ensure the robot marker can be represented in the output image
+	var robotMarker spatialmath.Pose
+	if ppRM.robotPose != nil {
+		robotMarker = *ppRM.robotPose
+		maxX = math.Max(maxX, robotMarker.Point().X)
+		minX = math.Min(minX, robotMarker.Point().X)
+		maxZ = math.Max(maxZ, robotMarker.Point().Z)
+		minZ = math.Min(minZ, robotMarker.Point().Z)
+	}
+
+	// Calculate the scale factors
+	widthScaleFactor := float64(imageWidth-1) / (maxX - minX)
+	heightScaleFactor := float64(imageHeight-1) / (maxZ - minZ)
+
+	// Add points in the pointcloud to a new image
+	im := rimage.NewImage(imageWidth, imageHeight)
+	cloud.Iterate(0, 0, func(pt r3.Vector, data pointcloud.Data) bool {
+		j := (pt.X - minX) * widthScaleFactor
+		i := (pt.Z - minZ) * heightScaleFactor
+		x, y := int(math.Round(j)), int(math.Round(i))
+
+		// Adds a point to an image using the value to define the color. If no value is available,
+		// the default color of white is used.
+		if x >= 0 && x < imageWidth && y >= 0 && y < imageHeight && data != nil {
+			var c rimage.Color
+			if data.HasValue() {
+				c = getProbabilityColorFromValue(data.Value())
+			} else {
+				c = rimage.NewColor(255, 255, 255)
+			}
+			im.Circle(image.Point{x, y}, voxelRadius, c)
+		}
+		return true
+	})
+
+	// Add a red robot marker to the image
+	if ppRM.robotPose != nil {
+		robotMarkerPoint := image.Point{
+			X: int(math.Round((robotMarker.Point().X - minX) * widthScaleFactor)),
+			Y: int(math.Round((robotMarker.Point().Z - minZ) * heightScaleFactor)),
```

Comment:
[minor] Same here, use the smaller scale factor for both axes

Reply:
Great call out!